dev/dev-env-setup.sh

```diff
@@ -0,0 +1,268 @@
+pyenv_exist=$(command -v pyenv)
+
+if [ -z "$pyenv_exist" ]; then
+  if [ -z "$GITHUB_ACTIONS" ]; then
+    read -p "pyenv is required to be installed to manage python versions. Would you like to install it? $(tput bold)(y/n)$(tput sgr0): " -n 1 -r
+    echo
+  fi
+  if [[ $REPLY =~ ^[Yy]$ || -n "$GITHUB_ACTIONS" ]]; then
+    if [[ "$machine" == mac ]]; then
+      # Check if brew is installed and install it if it isn't present
+      # Note: if xcode isn't installed, this will fail.
+      if [ -z "$(command -v brew)" ]; then
+        echo "Homebrew is required to install pyenv on MacOS. Installing in your home directory."
+        bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+      fi
+      echo "Updating brew and installing pyenv..."
+      echo "Note: this will probably take a considerable amount of time."
+      brew update
+      brew install pyenv
+      brew install openssl readline sqlite3 xz zlib
+    elif [[ "$machine" == linux ]]; then
+      sudo apt-get update -y
+      sudo apt-get install -y make build-essential libssl-dev zlib1g-dev \
+      libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
+      libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev
+      # Install pyenv from source
+      git clone --depth 1 https://github.com/pyenv/pyenv.git "$HOME/.pyenv"
+      PYENV_ROOT="$HOME/.pyenv"
+      PYENV_BIN="$PYENV_ROOT/bin"
+      PATH="$PYENV_BIN:$PATH"
+      if [ -n "$GITHUB_ACTIONS" ]; then
+        echo "$PYENV_BIN" >> "$GITHUB_PATH"
+        echo "PYENV_ROOT=$PYENV_ROOT" >> "$GITHUB_ENV"
+      fi
+    else
+      echo "Unknown operating system environment: $machine exiting."
+      exit 1
+    fi
+  else
+    PYENV_README=https://github.com/pyenv/pyenv/blob/master/README.md
+    echo "pyenv is required to use this environment setup script. Please install by following instructions here: $PYENV_README"
+    exit 1
+  fi
+fi
+
+MLFLOW_HOME=$(pwd)
+rd="$MLFLOW_HOME/requirements"
+
+# Get the minimum supported version from MLflow to ensure any feature development adheres to legacy Python versions
+min_py_version=$(python setup.py -q min_python_version)
+
+echo "The minimum version of Python to ensure backwards compatibility for MLflow development is: $(tput bold; tput setaf 3)$min_py_version$(tput sgr0)"
+
+if [[ -n "$override_py_ver" ]]; then
+  echo "$(tput bold; tput setaf 1)You are overriding the recommended version of Python for MLflow development: $min_py_version. $(tput sgr0)"
+  min_py_version="$(grep -o "^[0-9]*\.[0-9]*" <<< "$override_py_ver")"
+fi
+
+# Resolve a minor version to the latest micro version
+case $min_py_version in
+  "3.7") PY_INSTALL_VERSION="3.7.13" ;;
+  "3.8") PY_INSTALL_VERSION="3.8.13" ;;
+  "3.9") PY_INSTALL_VERSION="3.9.11" ;;
+  "3.10") PY_INSTALL_VERSION="3.10.3" ;;
+esac
+
+microver=$(grep -o '\.' <<< "$override_py_ver" | wc -l)
+if [[ $microver -gt 1 ]]; then
+  PY_INSTALL_VERSION=$override_py_ver
+fi
```

Comment:
The logic here is a bit difficult to understand to me:

```sh
minor_to_micro() {
  case $min_py_version in
    "3.7") echo "3.7.13" ;;
    "3.8") echo "3.8.13" ;;
    "3.9") echo "3.9.11" ;;
    "3.10") echo"3.10.3" ;;
  esac
}

if override_py_ver is specified
  if override_py_ver looks like minor version
      PY_INSTALL_VERSION=$(minor_to_micro $override_py_ver)
  elif override_py_ver looks like micro version
      PY_INSTALL_VERSION=$override_py_ver
  else
      # Invalid version
      exit 1
  fi
else
  PY_INSTALL_VERSION=$(minor_to_micro $min_py_version)
fi
```

Reply:
great idea! This logic is much easier to follow.